File: 76-minimum-window-substring/76-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        
        freq = [0]*256
        
        count = 0
        for c in t:
            if freq[ord(c) - ord("a")] == 0:
                count += 1
            freq[ord(c) - ord("a")] += 1
            
        
        l, r = 0, 0
        start, end = 0, 0
        res = 1000000
        
        while r < len(s):
            
            freq[ord(s[r]) - ord("a")] -= 1
            
            if freq[ord(s[r]) - ord("a")] == 0:
                count -= 1
                
            if count == 0:
                
                while count == 0:
                    if res > r-l+1:
                        res = r-l+1
                        start = l
                        end = start + r-l+1
                        
                    freq[ord(s[l]) - ord("a")] += 1
                    if freq[ord(s[l]) - ord("a")] == 1:
                        count += 1
                    l += 1
                    
            r += 1
        
        
        if res == 1000000:
            return ""
        
        return s[start : end]
        

File: 76-minimum-window-substring/76-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        
        int count = 0;
        int[] map = new int[256];
        
        for(int i=0; i<t.length(); i++){
            char ch = t.charAt(i);
            if(map[ch] == 0) count++;
            map[ch]++;
        }
        
        int i=0, j=0;
        int res = Integer.MAX_VALUE;
        int start = i;
        
        while(j < s.length()){
            
            char ch = s.charAt(j);
            
            map[ch]--;
            
            if(map[ch] == 0) count--;
            
            if(count == 0){
                
                while(count == 0){
                    
                    if(res > j-i+1){
                        res = j-i+1;
                        start = i;
                    }
                    
                    char c = s.charAt(i);
                    map[c]++;
                    if(map[c] == 1) count++;
                    
                    i++;
                }
            }
            
            j++;
        }
        
        if(res == Integer.MAX_VALUE) return "";
        return s.substring(start, start + res);
    }
}